File: packages/tanstack-workflow-core/src/server/parse-request.ts
// @ts-nocheck -- Vendored TanStack source targets a looser optional-property TypeScript policy.
/* oxlint-disable effect/restricted-syntax -- Vendored TanStack implementation source keeps upstream imperative control flow. */
import type { ApprovalResult, SignalDelivery } from '../types'

export interface WorkflowRequestParams {
  approval?: ApprovalResult
  /** Generic signal delivery. Mutually exclusive with `approval` in
   *  practice; `signalDelivery` takes precedence if both are set. */
  signalDelivery?: SignalDelivery
  input?: unknown
  runId?: string
  /** `true` when the client wants to cancel an in-flight run. */
  abort?: boolean
}

interface RawBody {
  abort?: boolean
  approval?: ApprovalResult
  signal?: SignalDelivery
  input?: unknown
  runId?: string
}

/**
 * Parse a workflow run request body. Returns params to spread into
 * `runWorkflow(...)`.
 *
 * @example
 * ```typescript
 * POST: async ({ request }) => {
 *   const params = await parseWorkflowRequest(request)
 *   if (params.abort && params.runId) {
 *     // ...host-specific abort plumbing
 *     return new Response(null, { status: 204 })
 *   }
 *   const stream = runWorkflow({ workflow, runStore, ...params })
 *   return toServerSentEventsResponse(stream)
 * }
 * ```
 */
export async function parseWorkflowRequest(
  request: Request,
): Promise<WorkflowRequestParams> {
  let raw: unknown
  try {
    raw = await request.json()
  } catch (err) {
    throw new WorkflowRequestParseError(
      err instanceof Error ? err.message : 'Invalid JSON body',
      err,
    )
  }
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
    throw new WorkflowRequestParseError(
      'Workflow request body must be a JSON object.',
    )
  }
  const body = raw as RawBody
  // Document precedence at the parse boundary: `signal` wins over
  // `approval` when both are set. The engine's resume path is
  // documented to ignore `approval` when `signalDelivery` is present,
  // but a forwarded `approval` next to `signalDelivery` is ambiguous
  // on the wire — normalize here so downstream code never has to
  // disambiguate.
  return {
    approval: body.signal ? undefined : body.approval,
    signalDelivery: body.signal,
    input: body.input,
    runId: body.runId,
    abort: body.abort,
  }
}

/**
 * Thrown by `parseWorkflowRequest` when the body cannot be parsed or
 * is not a JSON object. Route handlers should catch and return a 400.
 */
export class WorkflowRequestParseError extends Error {
  override readonly name = 'WorkflowRequestParseError'
  constructor(
    message: string,
    public override readonly cause?: unknown,
  ) {
    super(message)
  }
}

File: src/Firegrid.Core/ParseRequest.fs
namespace Firegrid.Core

open Fable.Core
open Fable.Core.JsInterop

/// Server request parsing (`server/parse-request.ts`).

/// Params to spread into `runWorkflow(...)`.
type WorkflowRequestParams =
    { Approval: ApprovalResult option
      SignalDelivery: SignalDelivery option
      Input: obj option
      RunId: string option
      Abort: bool option }

[<RequireQualifiedAccess>]
module ParseRequest =

    /// `WorkflowRequestParseError` JS class (name = "WorkflowRequestParseError").
    [<Emit("""(() => {
  const e = new Error($0);
  e.name = 'WorkflowRequestParseError';
  e._tag = 'WorkflowRequestParseError';
  if ($1 !== undefined) e.cause = $1;
  return e;
})()""")>]
    let private makeParseError (_message: string) (_cause: obj) : exn = jsNative

    let parseError (message: string) : exn = makeParseError message CoreSdk.undefinedValue

    let parseErrorWithCause (message: string) (cause: obj) : exn = makeParseError message cause

    let private optObj (o: obj) (key: string) : obj option =
        let v = CoreSdk.prop<obj> o key
        if CoreSdk.isUndefined v then None else Some v

    let private optStr (o: obj) (key: string) : string option =
        let v = CoreSdk.prop<obj> o key
        if CoreSdk.isNullish v then None else Some(CoreSdk.stringValue v)

    let private optBool (o: obj) (key: string) : bool option =
        let v = CoreSdk.prop<obj> o key
        if CoreSdk.isUndefined v then None else Some(unbox<bool> v)

    let private approvalOfObj (o: obj) : ApprovalResult =
        { Approved = CoreSdk.prop<bool> o "approved"
          ApprovalId = CoreSdk.prop<string> o "approvalId"
          Feedback = optStr o "feedback"
          Meta =
            let m = CoreSdk.prop<obj> o "meta"
            if CoreSdk.isNullish m then None else Some m }

    let private signalOfObj (o: obj) : SignalDelivery =
        { SignalId = CoreSdk.prop<string> o "signalId"
          StepId = optStr o "stepId"
          Name = CoreSdk.prop<string> o "name"
          Payload = CoreSdk.prop<obj> o "payload"
          Meta =
            let m = CoreSdk.prop<obj> o "meta"
            if CoreSdk.isNullish m then None else Some m }

    /// `parseWorkflowRequest(request)`.
    let parseWorkflowRequest (request: obj) : JS.Promise<WorkflowRequestParams> =
        promise {
            let! outcome = CoreSdk.requestJsonResult request

            let raw =
                if CoreSdk.prop<bool> outcome "ok" then
                    CoreSdk.prop<obj> outcome "value"
                else
                    let err = CoreSdk.prop<obj> outcome "error"
                    let message = if CoreSdk.isError err then CoreSdk.errorMessage err else "Invalid JSON body"
                    raise (parseErrorWithCause message err)

            // typeof raw !== "object" || raw === null || Array.isArray(raw)
            if not (CoreSdk.isTypeofObject raw) || isNull raw || CoreSdk.isArray raw then
                return raise (parseError "Workflow request body must be a JSON object.")
            else
                // `signal` wins over `approval`.
                let signal = optObj raw "signal"

                let signalDelivery = signal |> Option.map signalOfObj

                let approval =
                    match signal with
                    | Some _ -> None
                    | None -> optObj raw "approval" |> Option.map approvalOfObj

                return
                    { Approval = approval
                      SignalDelivery = signalDelivery
                      Input = optObj raw "input"
                      RunId = optStr raw "runId"
                      Abort = optBool raw "abort" }
        }
